File: backend-service/src/main/java/az/schedule/backendservice/service/TaskService.java
package az.schedule.backendservice.service;

import az.schedule.backendservice.dto.TaskDTO;
import az.schedule.backendservice.dto.request.task.AdvancedTaskFilterRequest;
import az.schedule.backendservice.dto.request.task.BulkTaskRequest;
import az.schedule.backendservice.dto.request.task.TaskRequest;
import az.schedule.backendservice.dto.response.BulkOperationResponse;
import az.schedule.backendservice.dto.response.PageResponse;
import az.schedule.backendservice.dto.response.TaskStatisticsResponse;
import az.schedule.backendservice.enums.Priority;
import az.schedule.backendservice.enums.TaskStatus;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public interface TaskService {
    TaskDTO createTask(TaskRequest request, Long accountId);
    
    TaskDTO updateTask(Long id, TaskRequest request, Long accountId);
    
    void deleteTask(Long id, Long accountId);
    
    TaskDTO getTaskById(Long id, Long accountId);
    
    PageResponse<TaskDTO> getTasksByAccount(Long accountId, Pageable pageable);
    
    PageResponse<TaskDTO> getTasksByAccountAndStatus(Long accountId, TaskStatus status, Pageable pageable);
    
    PageResponse<TaskDTO> getTasksByAccountAndCategory(Long accountId, Long categoryId, Pageable pageable);
    
    PageResponse<TaskDTO> searchTasks(Long accountId, String keyword, Pageable pageable);
    
    List<TaskDTO> getTasksByDateRange(Long accountId, LocalDateTime start, LocalDateTime end);
    
    TaskStatisticsResponse getAccountTaskStatistics(Long accountId);
    
    TaskStatisticsResponse getCategoryTaskStatistics(Long categoryId);
    
    // Bulk operations
    BulkOperationResponse bulkDeleteTasks(List<Long> taskIds, Long accountId);
    
    BulkOperationResponse bulkUpdateTaskStatus(BulkTaskRequest request, Long accountId);
    
    // Advanced filtering
    PageResponse<TaskDTO> advancedFilterTasks(AdvancedTaskFilterRequest filterRequest, Long accountId, Pageable pageable);
    
    // Priority filtering
    PageResponse<TaskDTO> getTasksByAccountAndPriority(Long accountId, Priority priority, Pageable pageable);
    
    // Overdue tasks
    List<TaskDTO> getOverdueTasks(Long accountId);
    
    long countTasksByAccount(Long accountId);
    
    long countTasksByAccountAndStatus(Long accountId, TaskStatus status);
    
    // Week and Month view
    List<TaskDTO> getTasksByWeek(Long accountId, LocalDateTime weekStart);
    
    List<TaskDTO> getTasksByMonth(Long accountId, int year, int month);
}

File: FE-AzSchedule/src/services/taskService.js
import axiosInstance from '../lib/axios';
import dayjs from 'dayjs';

/**
 * Create a new task
 * @param {Object} taskData - { title, description?, startTime?, endTime?, priority?, status, categoryId? }
 * @returns {Promise} Created task
 */
export const createTask = async (taskData) => {
  const response = await axiosInstance.post('/tasks', taskData);
  return response.data;
};

/**
 * Update a task
 * @param {number} id - Task ID
 * @param {Object} taskData - Update data
 * @returns {Promise} Updated task
 */
export const updateTask = async (id, taskData) => {
  const response = await axiosInstance.put(`/tasks/${id}`, taskData);
  return response.data;
};

/**
 * Delete a task
 * @param {number} id - Task ID
 * @returns {Promise} Delete response
 */
export const deleteTask = async (id) => {
  const response = await axiosInstance.delete(`/tasks/${id}`);
  return response.data;
};

/**
 * Get task by ID
 * @param {number} id - Task ID
 * @returns {Promise} Task data
 */
export const getTaskById = async (id) => {
  const response = await axiosInstance.get(`/tasks/${id}`);
  return response.data;
};

/**
 * Get all tasks with pagination
 * @param {Object} params - { pageNo?, pageSize?, sortBy?, sortDir? }
 * @returns {Promise} Paginated tasks
 */
export const getTasks = async (params = {}) => {
  const { pageNo = 0, pageSize = 10, sortBy = 'id', sortDir = 'desc' } = params;
  const response = await axiosInstance.get('/tasks', {
    params: { pageNo, pageSize, sortBy, sortDir }
  });
  return response.data;
};

/**
 * Get tasks by status
 * @param {string} status - Task status (TODO, IN_PROGRESS, DONE, PAUSED)
 * @param {Object} params - { pageNo?, pageSize?, sortBy?, sortDir? }
 * @returns {Promise} Paginated tasks
 */
export const getTasksByStatus = async (status, params = {}) => {
  const { pageNo = 0, pageSize = 10, sortBy = 'id', sortDir = 'desc' } = params;
  const response = await axiosInstance.get(`/tasks/status/${status}`, {
    params: { pageNo, pageSize, sortBy, sortDir }
  });
  return response.data;
};

/**
 * Get tasks by category
 * @param {number} categoryId - Category ID
 * @param {Object} params - { pageNo?, pageSize?, sortBy?, sortDir? }
 * @returns {Promise} Paginated tasks
 */
export const getTasksByCategory = async (categoryId, params = {}) => {
  const { pageNo = 0, pageSize = 10, sortBy = 'id', sortDir = 'desc' } = params;
  const response = await axiosInstance.get(`/tasks/category/${categoryId}`, {
    params: { pageNo, pageSize, sortBy, sortDir }
  });
  return response.data;
};

/**
 * Search tasks
 * @param {string} keyword - Search keyword
 * @param {Object} params - { pageNo?, pageSize?, sortBy?, sortDir? }
 * @returns {Promise} Paginated search results
 */
export const searchTasks = async (keyword, params = {}) => {
  const { pageNo = 0, pageSize = 10, sortBy = 'id', sortDir = 'desc' } = params;
  const response = await axiosInstance.get('/tasks/search', {
    params: { keyword, pageNo, pageSize, sortBy, sortDir }
  });
  return response.data;
};

/**
 * Get tasks by date range
 * @param {string} start - Start date time (ISO format)
 * @param {string} end - End date time (ISO format)
 * @returns {Promise} Tasks in date range
 */
export const getTasksByDateRange = async (start, end) => {
  const response = await axiosInstance.get('/tasks/date-range', {
    params: { start, end }
  });
  return response.data;
};

/**
 * Get task statistics for current user
 * @returns {Promise} Task statistics
 */
export const getTaskStatistics = async () => {
  const response = await axiosInstance.get('/tasks/statistics');
  return response.data;
};

/**
 * Get task statistics for a category
 * @param {number} categoryId - Category ID
 * @returns {Promise} Category task statistics
 */
export const getCategoryTaskStatistics = async (categoryId) => {
  const response = await axiosInstance.get(`/tasks/statistics/category/${categoryId}`);
  return response.data;
};

/**
 * Bulk delete tasks
 * @param {number[]} taskIds - Array of task IDs to delete
 * @returns {Promise} Bulk operation response
 */
export const bulkDeleteTasks = async (taskIds) => {
  const response = await axiosInstance.delete('/tasks/bulk', { data: taskIds });
  return response.data;
};

/**
 * Bulk update task status
 * @param {Object} data - { taskIds: number[], status: string }
 * @returns {Promise} Bulk operation response
 */
export const bulkUpdateTaskStatus = async (data) => {
  const response = await axiosInstance.post('/tasks/bulk/status', data);
  return response.data;
};

/**
 * Get tasks by priority
 * @param {string} priority - Priority level (LOW, MEDIUM, HIGH)
 * @param {Object} params - { pageNo?, pageSize?, sortBy?, sortDir? }
 * @returns {Promise} Paginated tasks
 */
export const getTasksByPriority = async (priority, params = {}) => {
  const { pageNo = 0, pageSize = 10, sortBy = 'id', sortDir = 'desc' } = params;
  const response = await axiosInstance.get(`/tasks/priority/${priority}`, {
    params: { pageNo, pageSize, sortBy, sortDir }
  });
  return response.data;
};

/**
 * Get overdue tasks
 * @returns {Promise} List of overdue tasks
 */
export const getOverdueTasks = async () => {
  const response = await axiosInstance.get('/tasks/overdue');
  return response.data;
};

/**
 * Advanced filter tasks
 * @param {Object} filterData - { statuses?, priorities?, categoryIds?, startDateFrom?, startDateTo?, endDateFrom?, endDateTo?, keyword? }
 * @param {Object} params - { pageNo?, pageSize?, sortBy?, sortDir? }
 * @returns {Promise} Filtered paginated tasks
 */
export const advancedFilterTasks = async (filterData, params = {}) => {
  const { pageNo = 0, pageSize = 10, sortBy = 'id', sortDir = 'desc' } = params;
  const response = await axiosInstance.post('/tasks/filter', filterData, {
    params: { pageNo, pageSize, sortBy, sortDir }
  });
  return response.data;
};

/**
 * Get tasks by week
 * @param {Date} weekStart - Week start date
 * @returns {Promise} List of tasks in the week
 */
export const getTasksByWeek = async (weekStart) => {
  // Format date as ISO string for backend
  const weekStartISO = dayjs(weekStart).format('YYYY-MM-DDTHH:mm:ss');
  const response = await axiosInstance.get('/tasks/week', {
    params: { weekStart: weekStartISO }
  });
  return response.data;
};

/**
 * Get tasks by month
 * @param {number} year - Year
 * @param {number} month - Month (1-12)
 * @returns {Promise} List of tasks in the month
 */
export const getTasksByMonth = async (year, month) => {
  const response = await axiosInstance.get('/tasks/month', {
    params: { year, month }
  });
  return response.data;
};
